include/mqtt/subscribe_options.hpp

```diff
@@ -0,0 +1,105 @@
+    subscribe_options(rap value) : data_(static_cast<std::uint8_t>(value)) { }
+    subscribe_options(nl value) : data_(static_cast<std::uint8_t>(value)) { }
+    subscribe_options(qos value) : data_(static_cast<std::uint8_t>(value)) { }
+
+    subscribe_options& operator|=(retain_handling value) { data_ |= static_cast<std::uint8_t>(value); return *this; }
+    subscribe_options& operator|=(rap value) { data_ |= static_cast<std::uint8_t>(value); return *this; }
+    subscribe_options& operator|=(nl value) { data_ |= static_cast<std::uint8_t>(value); return *this; }
+    subscribe_options& operator|=(qos value) { data_ |= static_cast<std::uint8_t>(value); return *this; }
+
+    retain_handling get_retain_handling() const
+    { return static_cast<retain_handling>(data_ & 0b00110000); }
+    rap get_rap() const
+    { return static_cast<rap>(data_ & 0b00001000); }
+    nl get_nl() const
+    { return static_cast<nl>(data_ & 0b00000100); }
+    qos get_qos() const
+    { return static_cast<qos>(data_ & 0b00000011); }
+
+    operator std::uint8_t() const { return data_; }
+    std::uint8_t data_;
+};
+
+subscribe_options operator|(subscribe_options lhs, retain_handling rhs) { lhs |= rhs; return lhs; }
+subscribe_options operator|(subscribe_options lhs, rap rhs) { lhs |= rhs; return lhs; }
+subscribe_options operator|(subscribe_options lhs, nl rhs) { lhs |= rhs; return lhs; }
+subscribe_options operator|(subscribe_options lhs, qos rhs) { lhs |= rhs; return lhs; }
+
+subscribe_options operator|(retain_handling lhs, rap rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(retain_handling lhs, nl rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(retain_handling lhs, qos rhs) { return subscribe_options(lhs) | rhs; }
+
+subscribe_options operator|(rap lhs, retain_handling rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(rap lhs, nl rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(rap lhs, qos rhs) { return subscribe_options(lhs) | rhs; }
+
+subscribe_options operator|(nl lhs, retain_handling rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(nl lhs, rap rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(nl lhs, qos rhs) { return subscribe_options(lhs) | rhs; }
+
+subscribe_options operator|(qos lhs, retain_handling rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(qos lhs, rap rhs) { return subscribe_options(lhs) | rhs; }
+subscribe_options operator|(qos lhs, nl rhs) { return subscribe_options(lhs) | rhs; }
+
+
+constexpr char const* qos_to_str(qos v) {
+    char const * const str[] = {
+        "at_most_once",
+        "at_least_once",
+        "exactly_once"
+    };
+    if (static_cast<std::uint8_t>(v) < sizeof(str)) return str[static_cast<std::uint8_t>(v)];
+    return "invalid_qos";
+}
+
+template<typename STREAM_T>
```

Comment:
```suggestion
template<typename Stream>
```

See https://github.com/redboltz/mqtt_cpp/wiki/Coding-Rules

Reply:
Fixed